### ISSUE TITLE
Additional specification around OTLP exporter retry requirements

### ISSUE THREAD

**Issue by jack-berg** [open-telemetry/opentelemetry-specification#1742]
**What are you trying to achieve?**

The OTLP exporter spec states the following concerning [retry logic](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/protocol/exporter.md#retry): 

> Transient errors MUST be handled with a retry strategy. This retry strategy MUST implement an exponential back-off with jitter to avoid overwhelming the destination until the network is restored or the destination has recovered.

There should be more specificity around how exactly the retry strategy should be configured with gRPC. [Here's the proto](https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L129-L151) definition for the gRPC Retry Policies. Without more specificity, language implementations may end up choosing arbitrary values for these parameters. 

Perhaps there should also be configuration parameters for adjusting the retry parameters. Maybe retry can be enabled / disabled altogether, or maybe the various parameters can be tuned. 

**Additional context.**

Discussing this in an opentelemetry-java [issue](https://github.com/open-telemetry/opentelemetry-java/issues/3271).

**Comment by tigrannajaryan** [comment]
> [Here's the proto](https://github.com/grpc/grpc-proto/blob/master/grpc/service_config/service_config.proto#L129-L151) definition for the gRPC Retry Policies.

This looks like a Message that describes a retry policy that can be serialized as a Protobuf, so I think it is unrelated to what this issue is about.

gRPC I believe has built-in interceptors that allow implementing exponential backoff with jitter: https://pkg.go.dev/github.com/grpc-ecosystem/go-grpc-middleware/retry#BackoffExponentialWithJitter

I do not think we need to mandate using this though. In the Collector for example we implement our [own backoff logic](https://github.com/open-telemetry/opentelemetry-collector/blob/0e3b991f9f2a5310ad60b26dfba8897d7bb94745/exporter/exporterhelper/queued_retry.go#L255) that is used by all exporters and is not gRPC specific.

>Without more specificity, language implementations may end up choosing arbitrary values for these parameters.
>Perhaps there should also be configuration parameters for adjusting the retry parameters. Maybe retry can be enabled / disabled altogether, or maybe the various parameters can be tuned.

Yeah, we should probably have recommendations for OTLP exporter implementations in the SDKs to follow.

**Comment by jack-berg** [comment]
> This looks like a Message that describes a retry policy that can be serialized as a Protobuf, so I think it is unrelated to what this issue is about.

I included the proto as it describes the schema of how you describe the retry policy when configuring a gRPC channel. [Here's](https://github.com/grpc/grpc/blob/master/doc/service_config.md) the general documentation for the service config mechanism. And [here's](https://github.com/grpc/proposal/blob/master/A6-client-retries.md) the documentation about the retry policy, which is a component of the service config. 

It appears that the built in gRPC retry mechanism is in proposal / experimental phase, but that its current status is that is being implemented for each language. So probably safe to start using it in the languages its available in.

> I do not think we need to mandate using this though. In the Collector for example we implement our own backoff logic that is used by all exporters and is not gRPC specific.

Sure. It seems like a good choice tho in languages where its available. And whatever recommendations are given for retry parameters should be compatible with the gRPC built in mechanism.

**Comment by inge4pres** [comment]
Let's revive this 😄 
Regardless of whether gRPC retries are implemented, the specification lacks a definition of how many and for how long individual requests should be re-tried.
This leaves implementation for the HTTP transport with guessing what they should do.

Is a more precise definition needed, or can we extend the text stating that authors can pick what they want?

**Comment by jpkrohling** [comment]
@jack-berg, before we mark this as "tc-inbox", perhaps you could share the current state around this?

**Comment by jack-berg** [comment]
> @jack-berg, before we mark this as "tc-inbox", perhaps you could share the current state around this?

It remains a problem but not one that is easily solvable. Back in 2021, I closed #1974 after failing to reach consensus. Since then, given the lack of specificity from the spec and the need to provide reliable components with retry enabled by default, various communities within otel have stabilized their own conception of how backoff should work (i.e. what parameters are configurable and the algo itself).

We could of course decide on a standard post-hoc, maybe based on one of the existing stable implementations, but no matter which implementation is chosen, there will be stable implementations which differ. 

In my experience as a opentelemetry-java maintainer, folks mostly just care that retry is toggleable and enabled by default. I haven't seen many (any?) users talk about tuning the parameters. There's are users interested in configuring which set of status codes are retriable, but we have a separate issue for that (#3876).

My advice for anyone looking to implement retry in an SDK is to follow the lead of the one of the existing implementations. 

@jpkrohling what would you have the TC do about this?

**Comment by danielgblanco** [comment]
@jack-berg is the TC recommendation not to specify aspects like how many times to retry, or the backoff strategy, and let language implementations handle it independently?

Follow up question, do you think that this sentence

> Transient errors MUST be handled with a retry strategy.

already specifies that retries must be enabled by default by any OTLP exporter?

If the answer is "yes" to both, we can close this issue.

**Comment by jpkrohling** [comment]
> @jpkrohling what would you have the TC do about this?

For triage purposes, I'd be interested in knowing if you think anything still needs to be done before the TC acts on this issue. @danielgblanco's questions go in this direction as well.
